### PR TITLE
[#5] Set up CI/CD with GitHub Actions

### LINKED ISSUE #5: [Chore] Set up CI/CD with Github Actions
## Why

App must have CI/CD pipeline to validate unit tests, release builds and showcase developer can setup pipeline

iOS deployment will be handled in #51 

## Acceptance Criteria

- [ ] Must run unit tests on every push
- [ ] Must build and push Android build to Firebase on merging release branch

### PR DESCRIPTION
**Pull request by ryan-conway**
- Closes https://github.com/nimblehq/flutter-ic-lydia-ryan/issues/5

## What happened 👀

- Added android staging deployment workflow
- Added iOS staging deployment workflow (Disabled)

## Insight 📝

Because iOS deployment isn't working right now, we have disabled the workflows in the repository for now and will revisit them in #51 

## Proof Of Work 📹

App was deployed to Firebase App distribution

Screenshot 2566-03-24 at 15 29 31

### CODE REVIEW

File: .github/workflows/ios_deploy_staging.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      - name: Bundle install
+        run: cd ./ios && bundle install
+
+      - name: Pod install
+        run: cd ./ios && pod install
+
+      - name: Deploy to Firebase App Distribution
+        run: cd ./ios && bundle exec fastlane build_and_upload_app
+        env:
+          FIREBASE_IOS_APP_ID_STAGING: ${{ secrets.FIREBASE_APP_ID_STAGING }}
```
Comment (lydiasama) [inline .github/workflows/ios_deploy_staging.yml]: Screenshot 2566-03-17 at 3 44 50 PM

@ryan-conway Could you please check the key here again? 😄  Is it supposed to be `FIREBASE_IOS_APP_ID_STAGING`?

File: ios/fastlane/Fastfile [hunk 2]
```diff
@@ -2,16 +2,10 @@
...
-require './Managers/DistributionManager'
 require './Managers/MatchManager'
 
-builder = BuildManager.new(fastlane: self)
-
-distribution_manager = DistributionManager.new(
-  fastlane: self,
-  build_path: Constants.BUILD_PATH
-)
+FIREBASE_APP_ID = ENV["FIREBASE_APP_ID_STAGING"]
```
Comment (lydiasama) [inline ios/fastlane/Fastfile]: Here too. Is it from Github secret?

Comment (ryan-conway) [reply ios/fastlane/Fastfile]: It's an environment variable, but the naming is still incorrect so we gotta update it too 🙈

File: .github/workflows/android_deploy_staging.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+name: Android - Deploy Staging build to Firebase
+
+on:
+  push:
+    branches:
+      - develop
```
Comment (lydiasama) [inline .github/workflows/android_deploy_staging.yml]: @ryan-conway Could you please set `chore/5-set-up-cicd-with-github-actions` for demo to get the deploy result for this PR? Then once we capture screen after app is uploaded, we can remove it.

**Comment by markgravity** [comment]
@ryan-conway Please set up to use Fastlane Match when deploying a build on the iOS side.
You can check this [PR](https://github.com/nimblehq/flutter-templates/pull/153) for example, it's still WIP but the Staging pipeline worked 🙏 
Please ping me on Slack if you need further help to set up it 🙋

File: .github/workflows/android_deploy_firebase.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+          echo $ENV_PRODUCTION > .env
+
+      - name: Build Android apk
+        run: flutter build apk --flavor production --debug --build-number $GITHUB_RUN_NUMBER
+
+      - name: Deploy Android to Firebase
+        uses: wzieba/Firebase-Distribution-Github-Action@v1.5.0
+        with:
+          appId: ${{ secrets.FIREBASE_ANDROID_APP_ID_PRODUCTION }}
+          token: ${{secrets.FIREBASE_TOKEN}}
```
Comment (lydiasama) [inline .github/workflows/android_deploy_firebase.yml:48]: @ryan-conway  Let me update this to the new approach: `serviceCredentialsFileContent`
Ref: https://github.com/wzieba/Firebase-Distribution-Github-Action#firebase-app-distribution-github-action

Comment (lydiasama) [reply .github/workflows/android_deploy_firebase.yml:48]: @ryan-conway It seems iOS template still use this `FIREBASE_TOKEN`. So I think we can leave it as-is to make all CD scripts work first. 🙈

File: ios/fastlane/Fastfile [hunk 5]
```diff
@@ -30,9 +33,14 @@ platform :ios do
...
 
-  desc 'Sync AppStore Staging match signing'
-  lane :sync_appstore_staging_signing do
-    match_manager.sync_app_store_signing(app_identifier: [Constants.BUNDLE_ID_STAGING])
+  desc 'Sync ad-hoc production match signing'
+  lane :sync_adhoc_signing do
+    match_manager.sync_adhoc_signing(app_identifier: [Constants.BUNDLE_ID_PRODUCTION])
+  end
+
+  desc 'Sync ad-hoc production match signing'
```
Comment (lydiasama) [inline ios/fastlane/Fastfile]:
```suggestion
  desc 'Sync AppStore production match signing'
```

File: ios/Runner/Info.plist [hunk 6]
```diff
@@ -22,7 +22,7 @@
 	<key>CFBundleSignature</key>
 	<string>????</string>
 	<key>CFBundleVersion</key>
-	<string>$(FLUTTER_BUILD_NUMBER)</string>
+	<string>4</string>
```
Comment (lydiasama) [inline ios/Runner/Info.plist]: @ryan-conway I'm not sure if we should revert this change? Is there `FLUTTER_BUILD_NUMBER` defined in somewhere?

File: ios/Runner.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -394,18 +394,21 @@
 				APP_DISPLAY_NAME = "Flutter Survey - Lydia Ryan";
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 				CLANG_ENABLE_MODULES = YES;
-				CURRENT_PROJECT_VERSION = "$(FLUTTER_BUILD_NUMBER)";
+				CODE_SIGN_IDENTITY = "Apple Development";
+				CODE_SIGN_STYLE = Automatic;
+				CURRENT_PROJECT_VERSION = 4;
```
Comment (lydiasama) [inline ios/Runner.xcodeproj/project.pbxproj]: Here too

File: .github/workflows/ios_deploy_to_testflight.yml [hunk 8]
```diff
@@ -1,50 +1,44 @@
...
     branches:
       - develop
 
 jobs:
-  build_and_upload_to_testflight:
-    name: Build And Upload iOS Application To TestFlight
-    runs-on: macOS-latest
+  build_and_upload_to_app_store:
+    name: Build and upload iOS app to AppStore
+    runs-on: macOS-12
```
Comment (lydiasama) [inline .github/workflows/ios_deploy_to_testflight.yml]: Can't we use the latest version of macOS? @ryan-conway 
Let's revert it for this PR. I think to revisit this workflow with the latest version is better. What do you think?

**Review by lydiasama — COMMENTED** [review]
A few concerns from my side.

**Review by lydiasama — APPROVED** [review]

File: .github/workflows/ios_deploy_firebase.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+jobs:
+  build_and_deploy_ios:
+    name: Build & Deploy iOS
+    runs-on: macOS-latest
+    environment: staging
+    env:
+      TEAM_ID: ${{ secrets.TEAM_ID }}
+      MATCH_PASSWORD: ${{ secrets.MATCH_PASSWORD }}
+      KEYCHAIN_PASSWORD: ${{ secrets.KEYCHAIN_PASSWORD }}
+      APPSTORE_CONNECT_API_KEY: ${{ secrets.APPSTORE_CONNECT_API_KEY }}
```
Comment (markgravity) [inline .github/workflows/ios_deploy_firebase.yml]: We will need this value when we upload a build to TestFlight or AppStore so it's a redundant value in this workflow 🙏

File: .github/workflows/ios_deploy_firebase.yml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: Install iOS dependencies
+        run: cd ./ios && bundle install && bundle exec pod install
+
+      - name: Match Adhoc
+        run: cd ./ios && bundle exec fastlane sync_adhoc_signing
+
+      - name: Deploy to Firebase App Distribution
+        run: cd ./ios && bundle exec fastlane build_and_upload_to_firebase
+        env:
```
Comment (markgravity) [inline .github/workflows/ios_deploy_firebase.yml:61]: Please setup to use `$GITHUB_RUN_NUMBER` as the build number 🙏 
Here is an example code for it
```
// ios_deploy_firebase.yml
- name: Deploy to Firebase App Distribution
         run: cd ./ios && bundle exec fastlane build_and_upload_to_firebase
         env:
         BUILD_NUMBER:  {{ $GITHUB_RUN_NUMBER }}

// Environment
  def self.BUILD_NUMBER
    ENV['BUILD_NUMBER']
  end

// Fastfile
desc 'Set build number '
private_lane :bump_build do |options|
  increment_build_number(
    build_number: Environments.BUILD_NUMBER,
    xcodeproj: Constants.PROJECT_PATH
  )
end
```

**Review by markgravity — COMMENTED** [review]
The rest LGTM

File: .github/workflows/ios_deploy_firebase.yml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+          FIREBASE_IOS_APP_ID: ${{ secrets.FIREBASE_IOS_APP_ID_PRODUCTION }}
+          FIREBASE_TOKEN: ${{ secrets.FIREBASE_TOKEN }}
+          BUILD_NUMBER:  {{ $GITHUB_RUN_NUMBER }}
```
Comment (markgravity) [inline .github/workflows/ios_deploy_firebase.yml:64]: Please help me to correct it 🙏 
`{{ $GITHUB_RUN_NUMBER }}` -> `${{ GITHUB_RUN_NUMBER }}`

By the way, can you add the POW for it to make sure it works as we expected please 🙏

File: .github/workflows/ios_deploy_to_testflight.yml [hunk 12]
```diff
@@ -1,50 +1,44 @@
...
 
     - name: Match AppStore
-      run: cd ./ios && bundle exec fastlane sync_appstore_staging_signing
+      run: cd ./ios && bundle exec fastlane sync_app_store_signing
 
-    - name: Deploy to TestFlight
+    - name: Deploy production build to TestFlight
       run: |
-        cd ./ios && bundle exec fastlane build_and_upload_testflight_app
+        cd ./ios && bundle exec fastlane build_and_upload_to_testflight
```
Comment (markgravity) [inline .github/workflows/ios_deploy_to_testflight.yml:44]: Please use the `BUILD_NUMBER` as well 🙏

File: .github/workflows/ios_deploy_firebase.yml (hunk 11 above)
Comment (ryan-conway) [reply .github/workflows/ios_deploy_firebase.yml:64]: @markgravity Currently we can't get the iOS deployment to work 🥲 We have decided to revisit this in the future so that we can focus on the Flutter development for now 🙏🏻 

[Slack convo](https://nimble-co.slack.com/archives/C04Q38BNSPL/p1679645899982609?thread_ts=1679638550.985039&cid=C04Q38BNSPL)

**Review by markgravity — APPROVED** [review]
